tests/components/hunterdouglas_powerview/test_config_flow.py

```diff
@@ -257,95 +130,121 @@ async def test_form_homekit_and_dhcp(
-            {"host": "1.2.3.4"},
+            {CONF_HOST: "1.2.3.4"},
         )
 
-    assert result2["type"] == "form"
+    assert result2["type"] == FlowResultType.FORM
     assert result2["errors"] == {"base": "cannot_connect"}
 
 
+@pytest.mark.usefixtures("mock_setup_entry", "mock_hunterdouglas_base")
 async def test_form_no_data(hass: HomeAssistant) -> None:
     """Test we handle no data being returned from the hub."""
     result = await hass.config_entries.flow.async_init(
         DOMAIN, context={"source": config_entries.SOURCE_USER}
     )
 
-    mock_powerview_userdata = _get_mock_powerview_userdata(userdata={"userData": {}})
     with patch(
-        "homeassistant.components.hunterdouglas_powerview.UserData",
-        return_value=mock_powerview_userdata,
+        "homeassistant.components.hunterdouglas_powerview.Hub.request_raw_data",
+        return_value={},
     ):
         result2 = await hass.config_entries.flow.async_configure(
             result["flow_id"],
-            {"host": "1.2.3.4"},
+            {CONF_HOST: "1.2.3.4"},
         )
 
-    assert result2["type"] == "form"
-    assert result2["errors"] == {"base": "unknown"}
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "cannot_connect"}
 
 
+@pytest.mark.usefixtures("mock_setup_entry", "mock_hunterdouglas_base")
 async def test_form_unknown_exception(hass: HomeAssistant) -> None:
     """Test we handle unknown exception."""
     result = await hass.config_entries.flow.async_init(
         DOMAIN, context={"source": config_entries.SOURCE_USER}
     )
 
-    mock_powerview_userdata = _get_mock_powerview_userdata(userdata={"userData": {}})
     with patch(
-        "homeassistant.components.hunterdouglas_powerview.UserData",
-        return_value=mock_powerview_userdata,
+        "homeassistant.components.hunterdouglas_powerview.config_flow.Hub.query_firmware",
+        side_effect=SyntaxError,
     ):
         result2 = await hass.config_entries.flow.async_configure(
             result["flow_id"],
-            {"host": "1.2.3.4"},
+            {CONF_HOST: "1.2.3.4"},
         )
 
-    assert result2["type"] == "form"
+    assert result2["type"] == FlowResultType.FORM
     assert result2["errors"] == {"base": "unknown"}
+
+
+@pytest.mark.usefixtures("mock_setup_entry", "mock_hunterdouglas_secondary")
+async def test_form_unsupported_device(hass: HomeAssistant) -> None:
+    """Test unsupported device failure."""
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {CONF_HOST: "1.2.3.4"},
+    )
+    await hass.async_block_till_done()
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "unsupported_device"}
```

Comment:
I just realized that some of the config flow tests aren't finishing the flows to reach either an ABORT or CREATE_ENTRY state.

All the config flow tests needs to show that it can recover from an error condition and reach either of the two final states.

Reply:
could you take a look at the ``async_step_user`` function for me before i go making changes incase i am on the wrong path

its still pretty much code from years ago and im thinking it also needs a block like the below
```
if error:
            return self.async_abort(reason=error)
```

It doesnt ever appear to return ``abort``, only a ``async_show_form`` - and all the tests you speak of appear to be mimicking the user_form and not a dicovery (discovery has abort)

Reply:
I pushed a sample in a181469d2e2832c69bdfd1ccf9712a951aa6b5b1